src/wiki/plugins/linknetwork/views.py

```diff
@@ -0,0 +1,127 @@
+from urllib.parse import quote as urlquote
+
+from django.contrib import messages
+from django.db.models import Q
+from django.http import HttpResponseForbidden
+from django.shortcuts import redirect
+from django.template.loader import render_to_string
+from django.utils.decorators import method_decorator
+from django.utils.translation import gettext as _
+from django.views.generic import ListView
+from django.views.generic import View
+from wiki import models as wiki_models
+from wiki.conf import settings
+from wiki.core.paginator import WikiPaginator
+from wiki.decorators import get_article
+from wiki.views.mixins import ArticleMixin
+
+from . import models
+
+
+def response_forbidden(request, article, urlpath, read_denied=False):
+    if request.user.is_anonymous:
+        qs = request.META.get("QUERY_STRING", "")
+        if qs:
+            qs = urlquote("?" + qs)
+        else:
+            qs = ""
+        return redirect(settings.LOGIN_URL + "?next=" + request.path + qs)
+    else:
+        return HttpResponseForbidden(
+            render_to_string(
+                "wiki/permission_denied.html",
+                context={
+                    "article": article,
+                    "read_denied": read_denied,
+                },
+                request=request,
+            )
+        )
+
+
+class WhatLinksHere(ListView, ArticleMixin):
+    template_name = "wiki/plugins/linknetwork/whatlinkshere.html"
+    allow_empty = True
+    context_object_name = "linknetwork"
+    paginator_class = WikiPaginator
+    paginate_by = 50
+    model = models.InternalLink
+
+    @method_decorator(get_article(can_read=True))
+    def dispatch(self, request, article, *args, **kwargs):
+        # If the article has been deleted, show a special page.
+        if article.current_revision and article.current_revision.deleted:
+            return redirect("wiki:deleted", article_id=article.id)
+
+        if not article.can_read(request.user):
+            return response_forbidden(request, article, read_denied=True)
+
+        return super().dispatch(request, article, *args, **kwargs)
+
+    def get_queryset(self):
+        # TODO: This filter should use query logic, instead of python filtering, as much as possible.
```

Comment:
Maybe there isn't much way around this! But you could try looking at the related manager, maybe `self.article.incoming_links.can_read(user=user)` can be used?

Minor thing related to another comment: What about deleted articles? Should they be filtered out as well? You can filter on something like `self.article.incoming_links.filter(current_revision__deleted=False).can_read(user=user)`.